### ISSUE TITLE
"'newChild == oldChild || !newChild.mounted' is not true" assertion

### ISSUE THREAD

**Issue by Hixie** [domokit/sky_engine#1177]
The following test gets an assertion. The code seems legit to me.

```dart
import 'package:sky/widgets.dart';
import 'package:test/test.dart';

import 'widget_tester.dart';

class Outer extends Component {
  Widget build() {
    return new Middle(new Inner());
  }
}

class Middle extends StatefulComponent {

  Middle(this.inner);

  Inner inner;

  void syncConstructorArguments(Outer source) {
    inner = source.inner;
  }

  Widget build() {
    return new SizeObserver(
      callback: _onSizeChanged,
      child: new Stack([
        inner,
      ])
    );
  }

  void _onSizeChanged(Size newSize) {
    setState(() { });
  }

}

class Inner extends StatefulComponent {

  void syncConstructorArguments(Inner source) { }

  Widget build() {
    return new SizeObserver(
      callback: _onSizeChanged,
      child: new Container()
    );
  }

  void _onSizeChanged(Size newSize) {
    setState(() { });
  }
}

void main() {
  test('setState() smoke test', () {
    WidgetTester tester = new WidgetTester();

    tester.pumpFrame(() {
      return new Outer();
    });

  });
}
```

cc @apwilson

**Comment by Hixie** [comment]
The problem (after I've removed the redundant assert and let syncChild() do the asserting instead of syncChildren()) is that when the Stack does its syncChildren, it calls syncChild() with a null old instead of an old==new situation.

**Comment by Hixie** [comment]
Oh, I see. It's because the lists are identical but two stateful components aren't _canSync()able.

**Comment by Hixie** [comment]
...and by the point we run syncChildren(), the child of the Stack is from a new generation, so we think it's been moved elsewhere and ignore it... hmm.

**Comment by Hixie** [comment]
Proposed fix in #1180.
